base/array.jl

```diff
@@ -2345,27 +2368,20 @@ julia> filter!(isodd, Vector(1:10))
 ```
 """
 function filter!(f, a::AbstractVector)
-    idx = eachindex(a)
-    y = iterate(idx)
-    y === nothing && return a
-    i, state = y
-
-    for acurr in a
-        if f(acurr)
-            @inbounds a[i] = acurr
-            y = iterate(idx, state)
-            y === nothing && (i += 1; break)
-            i, state = y
-        end
+    j = Int(firstindex(a))
+    for ai in a
+        @inbounds a[j] = ai
+        j = ifelse(f(ai), j+1, j)
+    end
+    if a isa Vector
+        resize!(a, j-1)
+        sizehint!(a, j-1)
+    else
+        Int(lastindex(a)) < j && deleteat!(a, j:Int(lastindex(a)))
```

Comment:
I don't know that `lastindex` is necessarily cast-able to an `Int`.

Reply:
Ok, I reworked the code to avoid arithmetic on potentially unknown index types. My reason for having the `Int`-cast was to ensure type-stability of `Int32`-indices.